
###################################
######### Path Configure ##########
###################################

# Path to metadata which contains at least sample name column and corresponding label column
META_PATH = '../dataset/metadata/mouse_sample_names_sra.tsv'

# Path to spatial transcriptomics imaging data
IMG_PATH = '../dataset/image/'

# Path to spatial transcriptomics gene counts data
CM_PATH = '../dataset/cm/'

# If spatial transcriptomics imaging data were aligned, leave it to None.
# Otherwise, give the path to affine transformation matrix generated by st_spot_detector
#  (ref: https://github.com/SpatialTranscriptomicsResearch/st_spot_detector)
ATM_PATH = None

# Path to folder that save the tiles
TILE_PATH = '../dataset/tile/'

# Path to save intermediate output and final result
DATASET_PATH = '../dataset/'

# Path to an image which will be used as a template for stain normalization
TEMPLATE_IMG = '../dataset/image/CN94_D2_HE.jpg'

####################################
######### Image Configure #########
####################################

# Tile size (DO NOT CHANGE)
SIZE = 299, 299
# Color channel (RGB)
N_CHANNEL = 3

# Image stain normalization method, can be
# 'macenko' (ref: M. Macenko et al., ‘A method for normalizing histology slides for quantitative analysis’,
#               in 2009 IEEE International Symposium on Biomedical Imaging: From Nano to Macro, 2009, pp. 1107–1110.)
# 'vahadane' (ref: A. Vahadane et al., ‘Structure-Preserving Color Normalization and Sparse Stain Separation for
#               Histological Images’, IEEE Transactions on Medical Imaging, vol. 35, no. 8, pp. 1962–1971, Aug. 2016.)
# 'reinhard' (ref: E. Reinhard, M. Adhikhmin, B. Gooch, and P. Shirley, ‘Color transfer between images’,
#               IEEE Computer Graphics and Applications, vol. 21, no. 5, pp. 34–41, Sep. 2001.)
NORM_METHOD = 'vahadane'

##########################################
######### Count Matrix Configure #########
##########################################

# Threshold for remove low abundant gene, genes expressed in less than THRESHOLD_GENE of total number spots  will be removed
THRESHOLD_GENE = 0.01

# Threshold for remove low quality spots, spots with less than THRESHOLD_SPOT of total genes expressed will be removed
THRESHOLD_SPOT = 0.01

# Minimum gene count value for counting whether expressed or not
# which are used for removing low abundant gene and low quality gene.
# genes with count value higher than MIN_EXP are considered as expressed genes
MIN_EXP = 1

######################################
######### Metadata configure #########
######################################

# Specify column name of sample name column, label column and
# condition column (used for subset if provided otherwise leave it to None) in metadata file
SAMPLE_COLUMN = 'sample_name'
LABEL_COLUMN = 'age'
CONDITION_COLUMN = 'breed'

# Subset dataset that all samples have certain CONDITION in CONDITION_COLUMN
CONDITION = 'B6SJLSOD1-G93A'
ADDITIONAL_COLUMN = 2 if CONDITION_COLUMN else 1

# Set random seed for reproducibility
seed = 37

# Color map for spots color in final clustering plot
color_map = ['#ff8aff', '#6fc23f', '#af63ff', '#eaed00', '#f02449', '#00dbeb', '#d19158', '#9eaada', '#89af7c', '#514036']

##################################################
######### Classification Model Configure #########
##################################################

# Option for run different model:
# "combine" : combine model that takes both image data and gene count data as input
# "gene_only" : single model that takes gene count data only as input
# "tile_only" : single model that takes images data only as input
model = ["combine", "gene_only", "tile_only"]

# Number of tiles that will be propagated through the model at each step
batch_size = 32

# Number of times that all training dataset will be passed forward and backward through model
epochs = 1

# The ratio for splitting training dataset and test dataset
train_ratio = 0.5

# Number of categories in label
n_classes = 4

##############################################
######### Cross Validation Configure #########
##############################################

# Option for stratified K-fold
# True : run stratified K-fold cross validation
# False : run model without cross validation
cross_validation = False

# Number of splits for cross validation
k_fold = 2
